---
layout:     post
title:      The comparison of different DE Analysis methods in microbiome sequencing data
subtitle:   Comparison of Type-I error rate and power
date:       2019-08-22
author:     Riki
header-img: img/microbiome.png
catalog: true
tags:
    - study
---

## INTRODUCTION
>_"A microbiota is an "ecological community of commensal, symbiotic and pathogenic microorganisms" found in and on all multicellular organisms studied to date from plants to animals. A microbiota includes bacteria, archaea, protists, fungi and viruses. Microbiota have been found to be crucial for immunologic, hormonal and metabolic homeostasis of their host. The synonymous term microbiome describes either the collective genomes of the microorganisms that reside in an environmental niche or the microorganisms themselves."_ (Wikipedia)

The microbiome represents all microbial species that are present in a particular environment. Advances in sequencing technology have facilitated the characterization of microbial communities in various environments and the identification of species that would be difficult, if not impossible, to cultivate through traditional methods. 

Alterations in microbiome composition have been linked to numerous health conditions, such as inflammatory bowel disease, obesity, and respiratory illnesses (Young 2017). In the context of human health, microbiome studies seek to understand the interaction between host biological and environmental factors and microbiome structure with the future goal of developing methods to treat disease through the modulation of microbial composition (Xia and Sun 2017). One method of assessing microbe-disease associations is through differential abundance testing, which identifies differences in mean abundances of various species between healthy and diseased individuals.

Common sequencing methods used to characterize microbiomes are marker gene surveys and whole genome shotgun sequencing. Marker gene surveys sequence a gene with high sequence variability between species, often the 16S rRNA gene, to identify microbes present within the sample; in contrast, whole genome sequencing does not have such restrictions. Reads with a specified amount of sequence similarity, typically ~97%, are clustered into operation taxonomic units (OTUs), which are representative of species (Jovel et al. 2016). The resulting data are organized into matrices with counts representing the abundance of each OTU in each sample.
	
Microbiome sequencing data is typically sparse with a high percentage of zero counts. Various sources have estimated this value to be approximately between 50-90% of the total sample counts (Weiss et al. 2017). These zeros may indicate true absence of the species in the sample or they may be caused by insufficient sequencing depth, an example of which is a rare species that is approaching the limit of detection by the machine. Additionally, the number of detected species is rarely saturated and further sequencing typically leads to more observed species (Weiss et al. 2017).

In this context, library size is defined as the total counts per sample. Differential sequencing efficiency can result in large differences in library sizes between samples, which can obscure true differences in species abundance. For example, if Sample A has five times the library size as that of Sample B, all of the OTUs in Sample A would appear to be differentially abundant even if the relative abundances between the two groups are similar, due to the fact that the absolute counts in Sample A would much higher. This library size effect can be reduced by using normalization methods, which remove biases and variation introduced by sampling and sequencing processes.

The goal of this study is to compare the performance of common normalization and differential abundance testing methods used for the analysis of microbiome data when sample library sizes are similar versus when they are different. The efficacy of these methods will be evaluated in terms of their type I error rates and power using data simulated from a zero-inflated negative binomial distribution. Previous studies have used existing microbiome datasets to compare methods in the presence of batch effects, but have not thoroughly addressed the library size issue (McMurdie and Holmes 2014; Pereira et al. 2018; Weiss et al. 2017).


## MATERIALS AND METHODS

We often want to cluster a batch of samples or find differential expressed OTUs between different groups. Due to the huge difference size, large amount of zeros and over-dispersion structure of microbiome sequencing data, it must be normalized. Otherwise, huge mistake may be caused.

- *Cluster*: Using certain normalization method + distance metrics (e.g.Bray-Curtis, Unifrac...)
- *Differential Analysis*:  Using certain differential abundance test, which contains a normalization preprocessing method.




## RESULTS

![Powertest_result_summary.png](https://i.loli.net/2019/08/21/49PQFLtcWmwDUuK.png)

This plot shows different traits of different test methods:
- **Wilcoxon**: Nearly no power at all due to the large amount of zeros(ties).
- **T-test**: Lower power than 
- **DESeq2**:  
- **edgeR**:  
- **CSS + MGS**: Have 

#### ZINB distribution

ZINB(Zero-Inflated Negative Binomial) distribution is often used to simulate OTU count data according to its zero-inflated and overdispersion features:

There are three parameters in a ZINB distribution:
- μ : mean of the non-ZI part
- θ : measure of dispersion
- π : the probability of the ZI part

We use a Gamma-Poisson mixture model to generate negative binomial variables:
```
rZINB <- function(n, param=NULL, m, t, p) 
  if (!is.null(param)) {m = param[1]; t = param[2]; p = param[3]}
  
  rvec <- rgamma(n, shape = m/t, rate = 1/t)
  rvec <- rpois(n, rvec)
  pvec <- rbinom(n, 1, 1-p)
  xvec <- rvec * pvec
  return(xvec)
}
```

After this Gamma-Poisson mixture process, we could get a random variable which has following distribution:  

![mRJszD.png](https://s2.ax1x.com/2019/08/26/mRJszD.png)

The parameter μ and θ can proportional reflects mean and variance of the random variable. For the non-ZI part (the simple NB part), we can calculate that:
- mean = μ
- variance = μ(θ+1)


#### Scenarios

First of all, we want to confirm our settings of the tested data:
- number of OTUs/taxa (rows of the matrix): 100
- number of samples (columns of the matrix): 2 groups * 50 samples/group = 100
- Each item of the matrix is generated from a ZINB distribution independently with the parameters: **μ = 10, θ = 5, π = 0.75**. These parameters were derived from an oral microbiome sequencing dataset from the ZOE2.0 study conducted at the University of North Carolina.

We designed three different scenarios of our generated data:
- **Scenario A**: Strictly designed same library size
- **Scenario B**: Completely random library size
- **Scenario C**: Mixed samples containing two different library sizes

To autualize these scenarios, we first generate a very large matrix with 1000000 (one million) samples, and then plot the histogram of library sizes of these one million samples:

![mhlKbR.png](https://s2.ax1x.com/2019/08/26/mhlKbR.png)



## DISCUSSION

## CONCLUSION


